### ISSUE TITLE
Rename build directory ('build' instead of 'bin')

### ISSUE THREAD

**Issue by straight-shoota** [crystal-lang/shards#179]
Currently, `shards build` places the build results in the `bin` folder. This folder is often used for tools and helper scripts in the repository (for example [Crystal comiler's `bin` dir](https://github.com/crystal-lang/crystal/tree/master/bin)).

It gets quite messy if this folder is simultaneously used for tools and as build directory. For example you can't can't just put `bin` in `.gitignore` or manually clean all build files with one `rm -rf bin`.

While it's certainly possible to put tools into a different folder (`scripts` is also quite popular), `bin` is commonly used for this. And I don't think shards should force people to use something different.

I suggest to use `build` instead of `bin` as a dedicated directory for build results. It makes great sense that `shards build` puts stuff in `build`.

**Comment by ysbaddaden** [comment]
Matter of taste and opinion. Mine is building a project binary MUST go into the `bin` folder, and tools may go to whatever (e.g. `tools` or `scripts`). Using something else makes no sense to me, and is highly unexpected.

We already force `lib` and `src`; I don't believe forcing `bin` is any more inconvenient.

**Comment by ysbaddaden** [comment]
BTW if a shell script and the binary would conflict, I always suffix the binary with `-binary`, and still out it inside `bin`.

**Comment by straight-shoota** [comment]
As with all naming issues, this is certainly a matter of taste. But I'm not convinced that we should force `bin` on everyone.
For `lib` and `src` it makes sense - to have a standardized layout to find things. But compiled binaries are generally not committed to a repo and not directly used by language tooling. So there is less encouragement for a need to standardize.

I'd appreciate it I you wouldn't simply discard any discussion about this. It's a matter of fact that there are different expectations and demands and not all are satisfied by the current behaviour. I strongly believe this deserves to open for discussion.

As I understand it, putting compiled binaries in `bin` folder comes from the C tradition and systems software. Java build tools Ant and Gradle use `build`, maven and Rust's Cargo `target`, Elixir's Mix `_build`. Judging from a quick look at Github's [.gitignore templates](https://github.com/github/gitignore) there are only a hand full of ignored `bin` folders, but quite a few `{.,_,}build`
One can hardly assume `bin` to be the only logical solution. The field is very diverse in that matter, and from the build tools I know or have looked at that decide on a default location for compiled binaries, in fact only `go build` uses `bin`. But as good as all of them agree to put source code in `src`.

Ruby obviously has no compiled binaries, but the `bin` folder is traditionally used for tools (like for the renowned `bin/setup` SOP). Most Crystal projects I know follow this custom, this includes the compiler (which puts compiled binaries in `.build`) and and even most of your shards 😉. While `scripts` or `tools` could certainly be used for this, `bin` is the de-facto standard in Crystalland.

Putting compiled binaries in the same directory is not just a matter of conflicting file names. I find `-binary` suffix a dirty solution btw., it's way better to simply put the build output in a separate location. If you have a dedicated folder for this, you can just put it in `.gitignore` to avoid checking in accidentally - or at least popping up as new files in `git status`.

**TL;DR** I think we should provide a sane default that makes it easy for everyone to use. As a matter of taste the exact name of the folder is really not important, but clearly `bin` is not a good target. Most Crystal projects put their tools there which leads to conflicts. It makes no sense to force everyone to change.

**Comment by RX14** [comment]
`lib` is used by shards and is gitngnored, how exactly is it different from `bin`?

**Comment by straight-shoota** [comment]
`bin` is quite often used for tools and scripts (like [`bin/setup`](https://robots.thoughtbot.com/bin-setup)) and therefore can't simply be gitignored.

**Comment by RX14** [comment]
Yes it can, I've seen `tools` and `scripts` in the wild just as many times as `bin`. Plus, when we implement shards being able to install project-lcoal tools, they'll go in `bin` too, and they should be gitignored.

**Comment by straight-shoota** [comment]
I haven't yet seen any Crystal project with anything else than `bin` as a tools folder. So yes, it is a weak convention, but I don't think it should simply be ignored and made unnecessary difficult for this most common case.

Tools should go in the tools folder. So the issue with gitignoring custom compiled tools exists whether this folder is called `bin`, `tools`, `scripts` or something else.

`shards install` can only install Crystal dependencies, so this can't be a reason to simplify things that are not Crystal.
For example a typical web application might need a few other things to setup like databases, migrations, sample content, message queues etc...

**Comment by RX14** [comment]
So it comes down to a matter of opinion. I doubt this argument is going to go anywhere.

**Comment by straight-shoota** [comment]
Yes it is based on opinion. According to my research, in Crystal projects the opinion tends to be using `bin` as a tools folder. Based on that knowledge I think it is not appropriate for `shards build` to use this as a build folder.

**Comment by ysbaddaden** [comment]
I heard your argument, but it's purely theoretical. Nobody seems annoyed in *practice*. Unless a *large* portion of the community believes this is incredibly bad, I won't reconsider.

**Comment by straight-shoota** [comment]
Please count at least me as being annoyed 😉
But fair enough, there is certainly no action required if it is just me. Let's see if there will be others...

**Comment by Blacksmoke16** [comment]
Has being able to customize the output location of built binaries been considered in the past?  Can of course have it default to `bin`.

My use case is related to the v2 Crystal snap plugin.  It expects that the built binaries are available at a specific location.  At the moment I'm doing `'cp -r ./bin "${SNAPCRAFT_PART_INSTALL}"/bin'`, but that feels kinda meh.  I've been referencing `Go` and `Rust` implementations and they have [GOBIN](https://medium.com/rungo/working-in-go-workspace-3b0576e0534a#:~:text=commands%20from%20here.-,GOBIN,-is%20another%20environment) and [--root](https://doc.rust-lang.org/cargo/commands/cargo-install.html#:~:text=Install%20all%20examples.-,--root,-dir) respectively.

**Comment by straight-shoota** [comment]
Maybe `shards build` could just accept and relay the `-o` option of `crystal build`.

It could even add the target name if the output path is a directory path (ends with separator).
